Does not build with GHC-9.0 due to TH failure

This package currently does not build with ghc 9.0 yet due to an upperbound on template-haskell. When I bumped the bound on template-haskell to 0.18 compilation fails with: 

```
src/Data/Vector/Circular.hs:330:15: error:
    • Couldn't match type: TExp a0
                     with: CircularVector a
      Expected: CircularVector a -> Code m (CircularVector a)
        Actual: CircularVector a -> Code m (TExp a0)
    • In the expression: unsafeTExpCoerce . lift
      In an equation for ‘liftTyped’: liftTyped = unsafeTExpCoerce . lift
      In the instance declaration for ‘Lift (CircularVector a)’
    • Relevant bindings include
        liftTyped :: CircularVector a -> Code m (CircularVector a)
          (bound at src/Data/Vector/Circular.hs:330:3)
    |
330 |   liftTyped = unsafeTExpCoerce . lift
    |               ^^^^^^^^^^^^^^^^^^^^^^^

src/Data/Vector/Circular.hs:560:3: error:
    • Couldn't match expected type: Q (TExp (CircularVector a))
                  with actual type: Code m0 (CircularVector a)
    • In the expression: liftTyped (unsafeFromList xs)
      In an equation for ‘vec’: vec xs = liftTyped (unsafeFromList xs)
    • Relevant bindings include
        xs :: [a] (bound at src/Data/Vector/Circular.hs:558:5)
        vec :: [a] -> Q (TExp (CircularVector a))
          (bound at src/Data/Vector/Circular.hs:557:1)
    |
560 |   liftTyped (unsafeFromList xs)
```
I might have a more detailed look on how to fix it later.

--- a/src/Data/Vector/Circular.hs
+++ b/src/Data/Vector/Circular.hs
@@ -326,9 +326,11 @@ instance Lift a => Lift (CircularVector a) where
     pure $ ConE ''CircularVector
       `AppE` (VarE 'NonEmpty.unsafeFromVector `AppE` v)
       `AppE` r
-#if MIN_VERSION_template_haskell(2,16,0)
+#if MIN_VERSION_template_haskell(2,17,0)
+  liftTyped = unsafeCodeCoerce . lift
+#elif MIN_VERSION_template_haskell(2,16,0)
   liftTyped = unsafeTExpCoerce . lift
-#endif /* MIN_VERSION_template_haskell(2,16,0) */
+#endif
 
 -- | Get the length of a 'CircularVector'.
 --
@@ -555,12 +557,7 @@ rotateLeft r' (CircularVector v r) = CircularVector v h
 --   @since 0.1
 vec :: Lift a => [a] -> Q (TExp (CircularVector a))
 vec [] = fail "Cannot create an empty CircularVector!"
-vec xs =
-#if MIN_VERSION_template_haskell(2,16,0)
-  liftTyped (unsafeFromList xs)
-#else
-  unsafeTExpCoerce [|unsafeFromList xs|]
-#endif /* MIN_VERSION_template_haskell(2,16,0) */
+vec xs = unsafeTExpCoerce (lift (unsafeFromList xs))
 
 -- | @since 0.1
 equivalent :: Ord a => CircularVector a -> CircularVector a -> Bool

--- a/src/Data/Vector/Circular/Generic.hs
+++ b/src/Data/Vector/Circular/Generic.hs
@@ -327,9 +327,11 @@ instance Lift a => Lift (CircularVector v a) where
     pure $ ConE ''CircularVector
       `AppE` v
       `AppE` r
-#if MIN_VERSION_template_haskell(2,16,0)
+#if MIN_VERSION_template_haskell(2,17,0)
+  liftTyped = unsafeCodeCoerce . lift
+#elif MIN_VERSION_template_haskell(2,16,0)
   liftTyped = unsafeTExpCoerce . lift
-#endif /* MIN_VERSION_template_haskell(2,16,0) */
+#endif
 
 -- | Get the length of a 'CircularVector'.
 --

--- a/vector-circular.cabal
+++ b/vector-circular.cabal
@@ -36,7 +36,7 @@ library
     , nonempty-vector >= 0.2 && < 0.3
     , primitive >= 0.6.4 && < 0.8
     , semigroupoids >= 5.3 && < 5.4
-    , template-haskell >= 2.12 && < 2.17
+    , template-haskell >= 2.12 && < 2.18
     , vector >= 0.12 && < 0.13
     , deepseq >= 1.4 && < 1.5
   ghc-options:

